File: Python/mergeSort.py
import math

def mergeSort(array):
    if(len(array)>1):
        m = len(array)//2
        L = array[:m]   #m not included
        R = array[m:]

        mergeSort(L)
        mergeSort(R)

        i=j=k=0

        while i < len(L) and j < len(R): 
            if L[i] < R[j]: 
                array[k] = L[i] 
                i+=1
            else: 
                array[k] = R[j] 
                j+=1
            k+=1
          
        # Checking if any element was left 
        while i < len(L): 
            array[k] = L[i] 
            i+=1
            k+=1
          
        while j < len(R): 
            array[k] = R[j] 
            j+=1
            k+=1


if __name__ == "__main__":
    print("\n------MERGE SORT------\n")
    n = int(input("Enter array size: "))

    array = []

    print("Insert array elements: ")
    for i in range(n):
        ele = int(input())
        array.append(ele)

    mergeSort(array)

    print("Sorted array: ")
    for i in range(n):
        print(array[i])

File: C++/mergeSort.cpp
#include<iostream>
#include<vector>
using namespace std;

void merge(int start, int mid, int end, vector<int> &arr) {
    int i = start, j = mid + 1, k = 0;

    vector<int> temp(end - start + 1, 0);

    while(i <= mid && j <= end) {
        if(arr[i] <= arr[j]) {
            temp[k] = arr[i];
            i += 1;
        } else {
            temp[k] = arr[j];
            j += 1;
        }
        k += 1;
    }


    while(i <= mid){
        temp[k] = arr[i];
        i += 1;
        k += 1;
    }

    while(j <= end){
        temp[k] = arr[j];
        j += 1;
        k += 1;
    }

    for(int x = start; x<=end; ++x){
        arr[x] = temp[x - start];
    }
    return;
}

void mergeSort(int start, int end, vector<int> &arr) {
    if(start >= end) return;

    int mid = (start + (end-start)/2);

    mergeSort(start, mid, arr);
    mergeSort(mid+1, end, arr);
    merge(start, mid, end, arr);
}

int main() {

    int n, ele;
    cout<<"Enter number of elements: ";
    cin>>n;

    vector<int> arr(n, 0);

    cout<<"Enter array values: \n";
    for(int i=0; i<n; ++i) {
        cin>>ele;
        arr[i] = ele;
    }

    cout<<"Initial array: ";
    for(int i=0; i<n; ++i) {
        cout<<arr[i]<<" ";
    }
    cout<<"\n";

    mergeSort(0, n-1, arr);

    cout<<"Sorted array: ";
    for(int i=0; i<n; ++i) {
        cout<<arr[i]<<" ";
    }
    cout<<"\n";
    
    return 0;
}